engines/dgds/decompress.cpp

```diff
@@ -0,0 +1,222 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/debug.h"
+#include "common/stream.h"
+#include "common/util.h"
+
+#include "dgds/decompress.h"
+
+namespace Dgds {
+
+uint32 RleDecompressor::decompress(byte *dest, uint32 sz, Common::SeekableReadStream &input) {
+	uint32 left = sz;
+
+	uint32 lenR = 0, lenW = 0;
+	while (left > 0 && !input.eos()) {
+		lenR = input.readByte();
+
+		if (lenR == 128) {
+			lenW = 0;
+		} else if (lenR <= 127) {
+			lenW = MIN(lenR, left);
+			for (uint32 j = 0; j < lenW; j++) {
+				*dest++ = input.readByte();
+			}
+			for (; lenR > lenW; lenR--) {
+				input.readByte();
+			}
+		} else {
+			lenW = MIN(lenR & 0x7F, left);
+			byte val = input.readByte();
+			memset(dest, val, lenW);
+			dest += lenW;
+		}
+
+		left -= lenW;
+	}
+
+	return sz - left;
+}
+
+void LzwDecompressor::reset() {
```

Comment:
maybe we can move the lzw decompressor into common code - this is probably the 10th implementation we have in engine code - but I'm not sure if they have subtle differences here and there.

Reply:
Agreed, it would be nice to combine them.  I also notice the RLE decoder in here is basically packbits and that is probably repeated multiple times too.  I'll take that as a TODO after merge if that's ok?

Reply:
Yes, this can be left as an overall TODO, as more engines will be affected by this merge